doc/reference/trd-non-xip.md

```diff
@@ -0,0 +1,94 @@
+Tock on Platforms Without Execute-in-Place (XIP) Flash
+========================================
+
+**TRD:** <br/>
+**Working Group:** Kernel<br/>
+**Type:** Informational<br/>
+**Status:** Draft <br/>
+**Author:** Brad Campbell<br/>
+**Draft-Created:** 2024/07/05<br/>
+**Draft-Modified:** 2024/07/05<br/>
+**Draft-Version:** 1<br/>
+**Draft-Discuss:** devel@lists.tockos.org<br/>
+
+Abstract
+-------------------------------
+
+This document provides information relating to running Tock on platforms without
+execute-in-place flash storage.
+
+Introduction
+===============================
+
+Traditional embedded systems use execute-in-place (XIP) flash to both store
+program code and to enable executing that program code. However, newer hardware
+designs, particularly those in security-focused settings, do not include any XIP
+flash storage. These platforms, subsequently referred to as "non-xip", only
+execute instructions stored in volatile RAM. All executable code must be stored
+in external nonvolatile storage, loaded into RAM, and then executed.
```

Comment:
```suggestion
execute instructions stored in volatile RAM. All executable code must be loaded
from external nonvolatile storage, loaded into RAM, and then executed.
```

Reply:
Is it not stored anywhere?

Reply:
This was very confusing to me during the discussion, the code has to exist _somewhere_, no? Or is part of the assumption that the machine code may actually be completely lost at some point?

Reply:
it's both true that it needs to be stored somewhere and that it needs to be loaded from somewhere. Between the two, I just think that the more salient part is that Tock needs to load it from somewhere.

Reply:
I think it is helpful to clarify that the assumption is the code is stored somewhere, and even if it is evicted from RAM it could be retrieved again. If that assumption is _not_ expected to be true I think that could very materially change potential Tock designs.

Reply:
I think that we shouldn't unnecessarily constrain ourselves by saying that Tock always has access to this code, which to me is implied by "stored". I think we want to support use-cases where a kernel has ephemeral access to binaries, or they are "streamed" over some interface (e.g., downloaded over TCP into memory). I think "loaded" encapsulates this a little better.

Reply:
In that case, that seems like a significant variation that we need to consider explicitly (and why this document exists).

Reply:
I'm not sure whether we talked about this _explicitly_ in the breakout session during TW7, but it felt implied. For instance, applications may be provided in storage that's "leased" from some larger chip where that memory is either not always available, or may be modified at any given time (for our purposes, volatile external memory). To support these use-cases it seems like it's vital for Tock's support to _not_ require unconditional access to such external storage.

But it'd be great if others can confirm whether or not that's a real concern. @lawrenceesswood @reynoldsbd

Reply:
I just want to replace the implication with something concretely written down one way or the other because I think my assumptions have been generally different from the others in that breakout.

Reply:
At work I often work on non-XIP MCU platforms, and one common use-case during development is the ability to stream code directly into RAM via JTAG and then execute from RAM, without having to store that code in flash at any intermediate point. This has several advantages:
1. it is faster
2. it reduces flash write cycles, which can be useful for hardware that is running hourly CI runs
3. if you have less flash than RAM, this enables running debug builds that only fit in RAM but do not fit in flash. This can be highly useful if your application is flash constrained in practice but you occasionally want to be able to run true debug builds (i.e. unoptimized code)

Of course, Tock could deign not to support this mode of operation. 

It is also possible on these platforms for OTA updates to overwrite the flash that contains the currently running application (since that application has already been copied into RAM). This is another common use case where Tock assuming that the application running in RAM matches the application that exists in flash could be problematic.